### PR TITLE
Feature/copy paste (#1436)

### PR DESCRIPTION
**Pull request by Raycoms**
It adds the possibility to paste messages directly into the world without requiring the builder.
For that, the player has to be in creative mode.
For now, it pastes it into the world with the correct building level + using the solid Substitution blocks.
In the future, we can add a second mode which turns of the solid substitution blocks.
It also fixes some bugs with the barracks.
And it fixes a bug where people lost their job after loading the barracks world.

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java [hunk 1]
```diff
@@ -240,35 +256,43 @@ public Object handle(
     public static class FlowerPotPlacementHandler implements IPlacementHandler
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java:256]: MAJOR Reduce this class from 39 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java [hunk 2]
```diff
@@ -65,39 +66,43 @@ private PlacementHandlers()
         handlers.add(new ChestPlacementHandler());
         handlers.add(new WayPointBlockPlacementHandler());
         handlers.add(new GeneralBlockPlacementHandler());
-
     }
 
     //If he woudln't count the bracket spaces we'd be under 25 easily.
     public static class FireplacementHandler implements IPlacementHandler
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java:72]: MAJOR Reduce this class from 36 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 3]
```diff
@@ -173,19 +173,32 @@ public void readFromNBT(@NotNull final NBTTagCompound compound)
...
-                    workers.add(data);
+                    final CitizenData data = getColony().getCitizen(workersTagList.getCompoundTagAt(i).getInteger(TAG_ID));
+                    if (data != null)
+                    {
+                        data.setWorkBuilding(this);
+                        workers.add(data);
+                    }
+                }
+            }
+            catch(final Exception e)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:189]: MAJOR Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityInfoPoster.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.TextComponentString;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+
+public class TileEntityInfoPoster extends TileEntity
+{
+    public final ITextComponent[] signText =
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityInfoPoster.java:15]: MINOR Make this final field static too. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1170)

File: src/main/java/com/minecolonies/coremod/util/StructureWrapper.java [hunk 5]
```diff
@@ -195,9 +210,10 @@ private boolean checkForFreeSpace(@NotNull final BlockPos pos)
     /**
      * Place a structure into the world.
      *
-     * @param pos coordinates
+     * @param pos       coordinates
+     * @param complete  paste it complete (with structure blocks) or without
      */
-    private void placeStructure(@NotNull final BlockPos pos)
+    private void placeStructure(@NotNull final BlockPos pos, final boolean complete)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/StructureWrapper.java:216]: MAJOR The Cyclomatic Complexity of this method "placeStructure" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Refactor this method to reduce its Cognitive Complexity from 41 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 95 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 44 major
* ![MINOR][MINOR] 45 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L137](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/structures/helpers/Structure.java#L137): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L247](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/structures/helpers/Structure.java#L247): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L260](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/structures/helpers/Structure.java#L260): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L36](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L36): class "WindowBuildTool" has 38 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L144](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L144): Remove this unused method parameter "stack". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L352](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L352): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.378 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L75](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L75): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 59 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L75](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L75): class "EntityCitizen" has 99 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L371](https://github.com/Minecolonies/minecolonies/blob/a08092b6624447494b41222e19c99162acf6b525/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L371): Extract this increment or decrement operator into a dedicated statement. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS881)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Asherslab** [comment]
Could you fix the few Codacy/Sonar errors? (ALSO could you suppress the three "Blocker" errors? they're getting annoying lol)

**Comment by Raycoms** [comment]
Yeah but not in this pr =)

**Comment by Raycoms** [comment]
We should create a separate pr for those blocking errors, the rest of the errors are sonar bugs.
We should eventually create a pr to ignore all those senseless issues.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 95 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 44 major
* ![MINOR][MINOR] 45 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L137](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/structures/helpers/Structure.java#L137): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L247](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/structures/helpers/Structure.java#L247): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L260](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/structures/helpers/Structure.java#L260): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L36](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L36): class "WindowBuildTool" has 38 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L144](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L144): Remove this unused method parameter "stack". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L352](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L352): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.378 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L75](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L75): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 59 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L75](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L75): class "EntityCitizen" has 99 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L371](https://github.com/Minecolonies/minecolonies/blob/6549d4df2bd865d1dfcbd260442e1a5570c3cd8f/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L371): Extract this increment or decrement operator into a dedicated statement. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS881)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**